Extract the certificate validation from main and use a context

matrixise opened the pull request.

matrixise committed:
Extract the certificate validation from main and use a context

JulienPalard commented:
Thanks for the PR @matrixise!

I'm not a huge fan of the `Context` object, I have a few ideas, feel free to tell me if you don't think it's better:

- Don't store `now()` in the context, just use `datetime.now()` every time, I know it's a context switch, but who cares?
- Don't store `verbose` in the context: use `logging.debug()` instead, and set loglevel to debug if `-v` is given.
- Raise in case of certificate error instead of returning a string.
- Give the limit as a `limit` parameter, as now the context only contains a limit.

I think it make it easier to use:

```python
print(validate_certificate(Context(now=datetime.utcnow(), limit=timedelta(days=14), verbose=args.verbose), Service("mdk.fr")))
```

to:

```python
try:
    validate_certificate(Service("mdk.fr"), limit=timedelta(days=14))
except CertificateValidationError as err:
    print(err)
```

what do you think?

matrixise committed:
Refactor via the comment of Julien

https://github.com/JulienPalard/certificate_watcher/pull/3#issuecomment-971684965

matrixise commented:
Question, when can we use the `args.verbose`, because I am not sure if this one is really useful.

matrixise committed:
Remove Context

JulienPalard commented:
I mostly used it while developping, now while demoing the app lib: it's not really interesting to see nothing, you can't tell if it works or not, so running with `--version` gives you all expirations dates, prooving the thing does something.

JulienPalard commented on certificate_watcher.py:
```diff
@@ -69,6 +70,28 @@ def __repr__(self):
+def validate_certificate(service: Service, limit: timedelta):
+    try:
+        cert = get_server_certificate(service)
+    except socket.timeout:
+        raise CertificateValidationError(f"{service}: connect timeout")
```
```suggestion
    except socket.timeout as err:
        raise CertificateValidationError(f"{service}: connect timeout") from err
```

Else Python will display "During the handling of the error, another error occured", which is misleading. With the raise it'll display something like "The exception is the cause of the other one", which is right.

JulienPalard commented on certificate_watcher.py:
```diff
@@ -69,6 +70,28 @@ def __repr__(self):
+        cert = get_server_certificate(service)
+    except socket.timeout:
+        raise CertificateValidationError(f"{service}: connect timeout")
+    except ConnectionResetError:
+        raise CertificateValidationError(f"{service}: Connection reset")
```
```suggestion
    except ConnectionResetError as err:
        raise CertificateValidationError(f"{service}: Connection reset") from err
```

JulienPalard commented on certificate_watcher.py:
```diff
@@ -69,6 +70,28 @@ def __repr__(self):
+        raise CertificateValidationError(f"{service}: connect timeout")
+    except ConnectionResetError:
+        raise CertificateValidationError(f"{service}: Connection reset")
+    except Exception as err:
+        raise CertificateValidationError(f"{service}: {err!s}")
```
```suggestion
    except Exception as err:
        raise CertificateValidationError(f"{service}: {err!s}") from err
```

JulienPalard commented:
I'm ~ok to drop --verbose, it's most very usefull, we can always add it back if needed, but if so please remove it from argparse.

matrixise committed:
Remove --verbose

matrixise committed:
Add the original exception in the stack

JulienPalard commented:
Thanks for the PR @matrixise !
